sympy/sets/sets.py

```diff
@@ -1009,34 +1010,36 @@ class Union(Set, EvalfMixin):
     """
     is_Union = True
 
-    def __new__(cls, *args, **kwargs):
-        evaluate = kwargs.get('evaluate', global_evaluate[0])
+    @property
+    def identity(self):
+        return S.EmptySet
 
-        # flatten inputs to merge intersections and iterables
-        args = list(args)
+    @property
+    def zero(self):
+        return S.UniversalSet
 
-        def flatten(arg):
-            if isinstance(arg, Set):
-                if arg.is_Union:
-                    return sum(map(flatten, arg.args), [])
-                else:
-                    return [arg]
-            if iterable(arg):  # and not isinstance(arg, Set) (implicit)
-                return sum(map(flatten, arg), [])
-            raise TypeError("Input must be Sets or iterables of Sets")
-        args = flatten(args)
+    def __new__(cls, *args, evaluate=None):
+        if evaluate is None:
+            evaluate = global_evaluate[0]
 
-        # Union of no sets is EmptySet
-        if len(args) == 0:
-            return S.EmptySet
+        # flatten inputs to merge intersections and iterables
+        args = _sympify(args)
 
         # Reduce sets using known rules
         if evaluate:
+            args = list(cls._new_args_filter(args))
             return simplify_union(args)
 
         args = list(ordered(args, Set._infimum_key))
 
-        return Basic.__new__(cls, *args)
+        obj = Basic.__new__(cls, *args)
+        obj._argset = frozenset(args)
+        return obj
+
+    @property
+    @cacheit
+    def args(self):
+        return self._args
```

Comment:
Should this be `self._argset?

Reply:
No, argset is a set. args needs to be a tuple. Without this property it uses the default on `LatticeOp`, which has a different ordering than the `_infimum_key` used by Union and Intersection.